From c84192e117e0e246b4d22b50c25c88b05cbc13c9 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 7 Feb 2025 17:16:57 +0100
Subject: [PATCH] bus: rifsc: support stm32mp21x debug entries

Add support for stm32mp21x platforms debug entry. The whole RIFSC IDs map
has been reworked. Also remove RIFSC data in generic
stm32_firewall_controller structure.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I9042207ce683d09cba8c4394e9df1e20184514e1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/437678
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/bus/stm32_rifsc.c          | 300 +++++++++++++++++++++++------
 include/linux/bus/stm32_firewall.h |   2 -
 2 files changed, 246 insertions(+), 56 deletions(-)

--- a/drivers/bus/stm32_rifsc.c
+++ b/drivers/bus/stm32_rifsc.c
@@ -74,11 +74,10 @@
 #define RIF_CID0			0x0
 #define RIF_CID1			0x1
 
-#ifdef CONFIG_DEBUG_FS
-#define STM32MP25_RIFSC_DEVICE_ENTRIES		128
-#define STM32MP25_RIFSC_MASTER_ENTRIES	16
-#define STM32MP25_RIFSC_RISAL_SUBREGIONS	2
-#define STM32MP25_RIFSC_RISAL_GRANULARITY	8
+#define RIFSC_DEVICE_ENTRIES		128
+#define RIFSC_INITIATOR_ENTRIES		16
+#define RIFSC_RISAL_SUBREGIONS		2
+#define RIFSC_RISAL_GRANULARITY		8
 
 #define RIFSC_RIMC_ATTR0		0xC10
 
@@ -87,15 +86,15 @@
 #define RIFSC_RIMC_MSEC			BIT(8)
 #define RIFSC_RIMC_MPRIV		BIT(9)
 
-#define RIFSC_RISC_SRCID_MASK	GENMASK(6, 4)
-#define RIFSC_RISC_SRPRIV	BIT(9)
-#define RIFSC_RISC_SRSEC	BIT(8)
-#define RIFSC_RISC_SRRLOCK	BIT(1)
-#define RIFSC_RISC_SREN	BIT(0)
+#define RIFSC_RISC_SRCID_MASK		GENMASK(6, 4)
+#define RIFSC_RISC_SRPRIV		BIT(9)
+#define RIFSC_RISC_SRSEC		BIT(8)
+#define RIFSC_RISC_SRRLOCK		BIT(1)
+#define RIFSC_RISC_SREN			BIT(0)
 #define RIFSC_RISC_SRLENGTH_MASK	GENMASK(27, 16)
-#define RIFSC_RISC_SRSTART_MASK	GENMASK(10, 0)
+#define RIFSC_RISC_SRSTART_MASK		GENMASK(10, 0)
 
-static const char *stm32mp25_rifsc_master_names[STM32MP25_RIFSC_MASTER_ENTRIES] = {
+static const char *stm32mp25_rifsc_initiator_names[RIFSC_INITIATOR_ENTRIES] = {
 	"ETR",
 	"SDMMC1",
 	"SDMMC2",
@@ -107,14 +106,33 @@ static const char *stm32mp25_rifsc_maste
 	"PCIE",
 	"GPU",
 	"DMCIPP",
-	"LTDC_L0/L1",
-	"LTDC_L2",
+	"LTDC_L1/L2",
+	"LTDC_L3",
 	"LTDC_ROT",
 	"VDEC",
 	"VENC"
 };
 
-static const char *stm32mp25_rifsc_dev_names[STM32MP25_RIFSC_DEVICE_ENTRIES] = {
+static const char *stm32mp21_rifsc_initiator_names[RIFSC_INITIATOR_ENTRIES] = {
+	"ETR",
+	"SDMMC1",
+	"SDMMC2",
+	"SDMMC3",
+	"OTG_HS",
+	"USBH",
+	"ETH1",
+	"ETH2",
+	"RESERVED",
+	"RESERVED",
+	"DCMIPP",
+	"LTDC_L1/L2",
+	"LTDC_L3",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+};
+
+static const char *stm32mp25_rifsc_dev_names[RIFSC_DEVICE_ENTRIES] = {
 	"TIM1",
 	"TIM2",
 	"TIM3",
@@ -234,8 +252,8 @@ static const char *stm32mp25_rifsc_dev_n
 	"I3C3",
 	"I3C4",
 	"ICACHE_DCACHE",
-	"LTDC_L0L1",
-	"LTDC_L2",
+	"LTDC_L1L2",
+	"LTDC_L3",
 	"LTDC_ROT",
 	"DSI_TRIG",
 	"DSI_RDFIFO",
@@ -245,7 +263,138 @@ static const char *stm32mp25_rifsc_dev_n
 	"IAC",
 };
 
-struct rifsc_dev_debug_data {
+static const char *stm32mp21_rifsc_dev_names[RIFSC_DEVICE_ENTRIES] = {
+	"TIM1",
+	"TIM2",
+	"TIM3",
+	"TIM4",
+	"TIM5",
+	"TIM6",
+	"TIM7",
+	"TIM8",
+	"TIM10",
+	"TIM11",
+	"TIM12",
+	"TIM13",
+	"TIM14",
+	"TIM15",
+	"TIM16",
+	"TIM17",
+	"RESERVED",
+	"LPTIM1",
+	"LPTIM2",
+	"LPTIM3",
+	"LPTIM4",
+	"LPTIM5",
+	"SPI1",
+	"SPI2",
+	"SPI3",
+	"SPI4",
+	"SPI5",
+	"SPI6",
+	"RESERVED",
+	"RESERVED",
+	"SPDIFRX",
+	"USART1",
+	"USART2",
+	"USART3",
+	"UART4",
+	"UART5",
+	"USART6",
+	"UART7",
+	"RESERVED",
+	"RESERVED",
+	"LPUART1",
+	"I2C1",
+	"I2C2",
+	"I2C3",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"SAI1",
+	"SAI2",
+	"SAI3",
+	"SAI4",
+	"RESERVED",
+	"MDF1",
+	"RESERVED",
+	"FDCAN",
+	"HDP",
+	"ADC1",
+	"ADC2",
+	"ETH1",
+	"ETH2",
+	"RESERVED",
+	"USBH",
+	"RESERVED",
+	"RESERVED",
+	"OTG_HS",
+	"DDRPERFM",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"STGEN",
+	"OCTOSPI1",
+	"RESERVED",
+	"SDMMC1",
+	"SDMMC2",
+	"SDMMC3",
+	"RESERVED",
+	"LTDC_CMN",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"CSI",
+	"DCMIPP",
+	"DCMI_PSSI",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RNG1",
+	"RNG2",
+	"PKA",
+	"SAES",
+	"HASH1",
+	"HASH2",
+	"CRYP1",
+	"CRYP2",
+	"IWDG1",
+	"IWDG2",
+	"IWDG3",
+	"IWDG4",
+	"WWDG1",
+	"RESERVED",
+	"VREFBUF",
+	"DTS",
+	"RAMCFG",
+	"CRC",
+	"SERC",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"I3C1",
+	"I3C2",
+	"I3C3",
+	"RESERVED",
+	"ICACHE_DCACHE",
+	"LTDC_L1L2",
+	"LTDC_L3",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"RESERVED",
+	"OTFDEC1",
+	"RESERVED",
+	"IAC",
+};
+
+struct stm32_rifsc_dev_debug_data {
 	char dev_name[15];
 	u8 dev_cid;
 	u8 dev_sem_cids;
@@ -256,7 +405,32 @@ struct rifsc_dev_debug_data {
 	bool dev_sec;
 };
 
-struct rifsc_master_debug_data {
+struct stm32_rifsc_resources_names {
+	const char **device_names;
+	const char **initiator_names;
+};
+
+static const struct stm32_rifsc_resources_names rifsc_mp21_res_names = {
+	.device_names = stm32mp21_rifsc_dev_names,
+	.initiator_names = stm32mp21_rifsc_initiator_names,
+};
+
+static const struct stm32_rifsc_resources_names rifsc_mp25_res_names = {
+	.device_names = stm32mp25_rifsc_dev_names,
+	.initiator_names = stm32mp25_rifsc_initiator_names,
+};
+
+struct rifsc_private {
+	const struct stm32_rifsc_resources_names *res_names;
+	u32 *risal_map_bases;
+	void __iomem *mmio;
+	unsigned int nb_risup;
+	unsigned int nb_rimu;
+	unsigned int nb_risal;
+};
+
+#ifdef CONFIG_DEBUG_FS
+struct rifsc_initiator_debug_data {
 	char m_name[11];
 	u8 m_cid;
 	bool cidsel;
@@ -274,22 +448,24 @@ struct rifsc_subreg_debug_data {
 	u16 sr_length;
 };
 
-static void stm32_rifsc_fill_master_dbg_entry(struct stm32_firewall_controller *rifsc,
-					      struct rifsc_master_debug_data *dbg_entry, int i)
+static void stm32_rifsc_fill_initiator_dbg_entry(struct rifsc_private *rifsc,
+						 struct rifsc_initiator_debug_data *dbg_entry,
+						 int i)
 {
+	const struct stm32_rifsc_resources_names *dbg_names = rifsc->res_names;
 	u32 rimc_attr = readl_relaxed(rifsc->mmio + RIFSC_RIMC_ATTR0 + 0x4 * i);
 
-	snprintf(dbg_entry->m_name, sizeof(dbg_entry->m_name), "%s",
-		 stm32mp25_rifsc_master_names[i]);
+	snprintf(dbg_entry->m_name, sizeof(dbg_entry->m_name), "%s", dbg_names->initiator_names[i]);
 	dbg_entry->m_cid = FIELD_GET(RIFSC_RIMC_MCID_MASK, rimc_attr);
 	dbg_entry->cidsel = rimc_attr & RIFSC_RIMC_CIDSEL;
 	dbg_entry->m_sec = rimc_attr & RIFSC_RIMC_MSEC;
 	dbg_entry->m_priv = rimc_attr & RIFSC_RIMC_MPRIV;
 }
 
-static void stm32_rifsc_fill_dev_dbg_entry(struct stm32_firewall_controller *rifsc,
-					   struct rifsc_dev_debug_data *dbg_entry, int i)
+static void stm32_rifsc_fill_dev_dbg_entry(struct rifsc_private *rifsc,
+					   struct stm32_rifsc_dev_debug_data *dbg_entry, int i)
 {
+	const struct stm32_rifsc_resources_names *dbg_names = rifsc->res_names;
 	u32 cid_cfgr, sec_cfgr, priv_cfgr;
 	u8 reg_id = i / IDS_PER_RISC_SEC_PRIV_REGS;
 	u8 reg_offset = i % IDS_PER_RISC_SEC_PRIV_REGS;
@@ -299,7 +475,7 @@ static void stm32_rifsc_fill_dev_dbg_ent
 	priv_cfgr = readl_relaxed(rifsc->mmio + RIFSC_RISC_PRIVCFGR0 + 0x4 * reg_id);
 
 	snprintf(dbg_entry->dev_name, sizeof(dbg_entry->dev_name), "%s",
-		 stm32mp25_rifsc_dev_names[i]);
+		 dbg_names->device_names[i]);
 	dbg_entry->dev_id = i;
 	dbg_entry->dev_cid_filt_en = cid_cfgr & CIDCFGR_CFEN;
 	dbg_entry->dev_sem_en = cid_cfgr & CIDCFGR_SEMEN;
@@ -309,9 +485,9 @@ static void stm32_rifsc_fill_dev_dbg_ent
 	dbg_entry->dev_priv = priv_cfgr & BIT(reg_offset) ?  true : false;
 }
 
-static void stm32_rifsc_fill_subreg_dbg_entry(struct stm32_firewall_controller *rifsc,
+static void stm32_rifsc_fill_subreg_dbg_entry(struct rifsc_private *rifsc,
 					      struct rifsc_subreg_debug_data *dbg_entry, int i,
-						  int j)
+					      int j)
 {
 	u32 risc_xcfgr = readl_relaxed(rifsc->mmio + RIFSC_RISC_REG0_ACFGR + 0x10 * i + 0x8 * j);
 	u32 risc_xaddr;
@@ -333,7 +509,7 @@ static void stm32_rifsc_fill_subreg_dbg_
 
 static int stm32_rifsc_conf_dump_show(struct seq_file *s, void *data)
 {
-	struct stm32_firewall_controller *rifsc = (struct stm32_firewall_controller *)s->private;
+	struct rifsc_private *rifsc = (struct rifsc_private *)s->private;
 	u32 subregion_start, subregion_end;
 	int i, j;
 
@@ -354,8 +530,8 @@ static int stm32_rifsc_conf_dump_show(st
 	seq_puts(s, "| SCID |");
 	seq_printf(s, "| %7s |\n", "SEMWL");
 
-	for (i = 0; i < STM32MP25_RIFSC_DEVICE_ENTRIES; i++) {
-		struct rifsc_dev_debug_data d_dbg_entry;
+	for (i = 0; i < RIFSC_DEVICE_ENTRIES && i < rifsc->nb_risup; i++) {
+		struct stm32_rifsc_dev_debug_data d_dbg_entry;
 
 		stm32_rifsc_fill_dev_dbg_entry(rifsc, &d_dbg_entry, i);
 
@@ -375,18 +551,18 @@ static int stm32_rifsc_conf_dump_show(st
 	seq_puts(s, "                  RIMU dump\n");
 	seq_puts(s, "=============================================\n");
 
-	seq_puts(s, "| Master name |");
+	seq_puts(s, "| Initiator name |");
 	seq_puts(s, "| CIDSEL |");
 	seq_puts(s, "| MCID |");
 	seq_puts(s, "| N/SECURE |");
 	seq_puts(s, "| N/PRIVILEGED |\n");
 
-	for (i = 0; i < STM32MP25_RIFSC_MASTER_ENTRIES; i++) {
-		struct rifsc_master_debug_data m_dbg_entry;
+	for (i = 0; i < RIFSC_INITIATOR_ENTRIES && rifsc->nb_rimu; i++) {
+		struct rifsc_initiator_debug_data m_dbg_entry;
 
-		stm32_rifsc_fill_master_dbg_entry(rifsc, &m_dbg_entry, i);
+		stm32_rifsc_fill_initiator_dbg_entry(rifsc, &m_dbg_entry, i);
 
-		seq_printf(s, "| %-11s |", m_dbg_entry.m_name);
+		seq_printf(s, "| %-14s |", m_dbg_entry.m_name);
 		seq_printf(s, "| %-6s |", m_dbg_entry.cidsel ? "CIDSEL" : "");
 		seq_printf(s, "| %-4d |", m_dbg_entry.m_cid);
 		seq_printf(s, "| %-8s |", m_dbg_entry.m_sec ? "SEC" : "NSEC");
@@ -409,7 +585,7 @@ static int stm32_rifsc_conf_dump_show(st
 		seq_puts(s, "|  Subreg. end  |\n");
 
 		for (i = 0; i < rifsc->nb_risal; i++) {
-			for (j = 0; j < STM32MP25_RIFSC_RISAL_SUBREGIONS; j++) {
+			for (j = 0; j < RIFSC_RISAL_SUBREGIONS; j++) {
 				struct rifsc_subreg_debug_data sr_dbg_entry;
 
 				stm32_rifsc_fill_subreg_dbg_entry(rifsc, &sr_dbg_entry, i, j);
@@ -425,12 +601,11 @@ static int stm32_rifsc_conf_dump_show(st
 					   sr_dbg_entry.sr_enable ? "enabled" : "disabled");
 
 				subregion_start = rifsc->risal_map_bases[2 * i] +
-					sr_dbg_entry.sr_start * STM32MP25_RIFSC_RISAL_GRANULARITY;
+					sr_dbg_entry.sr_start * RIFSC_RISAL_GRANULARITY;
 				subregion_end = rifsc->risal_map_bases[2 * i] +
 					rifsc->risal_map_bases[2 * i + 1] - 1;
 				subregion_end = min(subregion_start + sr_dbg_entry.sr_length *
-						    STM32MP25_RIFSC_RISAL_GRANULARITY - 1,
-						    subregion_end);
+						    RIFSC_RISAL_GRANULARITY - 1, subregion_end);
 
 				seq_printf(s, "| 0x%-11x |", subregion_start);
 				seq_printf(s, "| 0x%-11x |\n", subregion_end);
@@ -442,7 +617,7 @@ static int stm32_rifsc_conf_dump_show(st
 }
 DEFINE_SHOW_ATTRIBUTE(stm32_rifsc_conf_dump);
 
-static int stm32_rifsc_register_debugfs(struct stm32_firewall_controller *controller)
+static int stm32_rifsc_register_debugfs(struct rifsc_private *priv)
 {
 	struct dentry *root = NULL;
 
@@ -453,12 +628,12 @@ static int stm32_rifsc_register_debugfs(
 	if (IS_ERR(root))
 		return PTR_ERR(root);
 
-	debugfs_create_file("rifsc", 0444, root, controller, &stm32_rifsc_conf_dump_fops);
+	debugfs_create_file("rifsc", 0444, root, priv, &stm32_rifsc_conf_dump_fops);
 
 	return 0;
 }
 #else /* CONFIG_DEBUG_FS */
-static int stm32_rifsc_register_debugfs(struct stm32_firewall_controller *controller __unused)
+static int stm32_rifsc_register_debugfs(struct rifsc_private *priv __unused)
 {
 	return 0;
 }
@@ -652,6 +827,7 @@ static int stm32_rifsc_populate_bus(stru
 static int stm32_rifsc_probe(struct platform_device *pdev)
 {
 	struct stm32_firewall_controller *rifsc_controller;
+	struct rifsc_private *rifsc_priv;
 	struct device_node *np = pdev->dev.of_node;
 	u32 nb_risup, nb_rimu, nb_risal;
 	struct resource *res;
@@ -662,6 +838,10 @@ static int stm32_rifsc_probe(struct plat
 	if (!rifsc_controller)
 		return -ENOMEM;
 
+	rifsc_priv = devm_kzalloc(&pdev->dev, sizeof(*rifsc_priv), GFP_KERNEL);
+	if (!rifsc_priv)
+		return -ENOMEM;
+
 	mmio = devm_platform_get_and_ioremap_resource(pdev, 0, &res);
 	if (IS_ERR(mmio))
 		return PTR_ERR(mmio);
@@ -681,8 +861,14 @@ static int stm32_rifsc_probe(struct plat
 	nb_risal = FIELD_GET(HWCFGR2_CONF3_MASK,
 			     readl(rifsc_controller->mmio + RIFSC_RISC_HWCFGR2));
 	rifsc_controller->max_entries = nb_risup + nb_rimu + nb_risal;
-	rifsc_controller->nb_risal = nb_risal;
 
+	rifsc_priv->nb_rimu = nb_rimu;
+	rifsc_priv->nb_risup = nb_risup;
+	rifsc_priv->nb_risal = nb_risal;
+	rifsc_priv->mmio = mmio;
+	rifsc_priv->res_names = of_device_get_match_data(&pdev->dev);
+	if (!rifsc_priv->res_names)
+		return -ENODEV;
 	/*
 	 * In STM32MP21, RIFSC_RISC_HWCFGR2 shows an incorrect number of RISAL (NUM_RISAL is 3
 	 * instead of 0). A software workaround is implemented using the st,mem-map property in the
@@ -699,22 +885,21 @@ static int stm32_rifsc_probe(struct plat
 	}
 
 	if (nb_risal_map_bases == 0)
-		rifsc_controller->nb_risal = 0;
+		rifsc_priv->nb_risal = 0;
 
 	/* Get RISAL map bases */
-	if (rifsc_controller->nb_risal > 0) {
-		if (nb_risal_map_bases != 2 * rifsc_controller->nb_risal) {
+	if (rifsc_priv->nb_risal > 0) {
+		if (nb_risal_map_bases != 2 * rifsc_priv->nb_risal) {
 			pr_err("RISAL count in HW configuration register and device tree mismatch");
 			return -EINVAL;
 		}
 
-		rifsc_controller->risal_map_bases = devm_kzalloc(&pdev->dev, nb_risal_map_bases *
-								sizeof(u32), GFP_KERNEL);
-		if (!rifsc_controller->risal_map_bases)
+		rifsc_priv->risal_map_bases = devm_kzalloc(&pdev->dev, nb_risal_map_bases *
+							   sizeof(u32), GFP_KERNEL);
+		if (!rifsc_priv->risal_map_bases)
 			return -ENOMEM;
 
-		err = of_property_read_u32_array(np, "st,mem-map",
-						 rifsc_controller->risal_map_bases,
+		err = of_property_read_u32_array(np, "st,mem-map", rifsc_priv->risal_map_bases,
 						 nb_risal_map_bases);
 		if (err) {
 			pr_err("Couldn't read st,mem-map property");
@@ -738,7 +923,7 @@ static int stm32_rifsc_probe(struct plat
 		return rc;
 	}
 
-	rc = stm32_rifsc_register_debugfs(rifsc_controller);
+	rc = stm32_rifsc_register_debugfs(rifsc_priv);
 	if (rc) {
 		dev_err(rifsc_controller->dev, "Couldn't add RIFSC debug entry: %d", rc);
 		return rc;
@@ -749,7 +934,14 @@ static int stm32_rifsc_probe(struct plat
 }
 
 static const struct of_device_id stm32_rifsc_of_match[] = {
-	{ .compatible = "st,stm32mp25-rifsc" },
+	{
+		.compatible = "st,stm32mp25-rifsc",
+		.data = &rifsc_mp25_res_names,
+	},
+	{
+		.compatible = "st,stm32mp21-rifsc",
+		.data = &rifsc_mp21_res_names,
+	},
 	{}
 };
 MODULE_DEVICE_TABLE(of, stm32_rifsc_of_match);
--- a/include/linux/bus/stm32_firewall.h
+++ b/include/linux/bus/stm32_firewall.h
@@ -45,8 +45,6 @@ struct stm32_firewall_controller {
 	struct list_head entry;
 	unsigned int type;
 	unsigned int max_entries;
-	u32 *risal_map_bases;
-	u32 nb_risal;
 
 	int (*grant_access)(struct stm32_firewall_controller *ctrl, u32 id);
 	void (*release_access)(struct stm32_firewall_controller *ctrl, u32 id);
